components/script/dom/htmlvideoelement.rs

```diff
@@ -416,3 +447,56 @@ impl PosterFrameFetchContext {
         }
     }
 }
+
+pub trait LayoutHTMLVideoElementHelpers {
+    fn data(self) -> HTMLMediaData;
+    fn get_width(self) -> LengthOrPercentageOrAuto;
+    fn get_height(self) -> LengthOrPercentageOrAuto;
+}
+
+impl LayoutDom<'_, HTMLVideoElement> {
+    fn width_attr(self) -> Option<LengthOrPercentageOrAuto> {
+        self.upcast::<Element>()
+            .get_attr_for_layout(&ns!(), &local_name!("width"))
+            .map(AttrValue::as_dimension)
+            .cloned()
+    }
+
+    fn height_attr(self) -> Option<LengthOrPercentageOrAuto> {
+        self.upcast::<Element>()
+            .get_attr_for_layout(&ns!(), &local_name!("height"))
+            .map(AttrValue::as_dimension)
+            .cloned()
+    }
+}
+
+impl LayoutHTMLVideoElementHelpers for LayoutDom<'_, HTMLVideoElement> {
+    fn data(self) -> HTMLMediaData {
+        let video = self.unsafe_get();
+
+        // Get the current frame being rendered.
+        let current_frame = video.htmlmediaelement.get_current_frame_data();
+
+        // This value represents the natural width and height of the video.
+        // It may exist even if there is no current frame (for example, after the
+        // metadata of the video is loaded).
+        let metadata = video
+            .get_video_width()
+            .zip(video.get_video_height())
+            .map(|(width, height)| MediaMetadata { width, height });
+
+        HTMLMediaData {
+            current_frame,
+            metadata,
+            has_default_size: self.width_attr().is_none() || self.height_attr().is_none(),
```

Comment:
I don't think this should depend on the attributes, e.g.

```html
<video width="100" height="100" style="border: solid; height: auto; aspect-ratio: auto"></video>
```

should still use the default aspect ratio of 2.

So  would just remove  `has_default_size` and handle it during layout.

Reply:
I think the video element has a special case for this, which is what the `video-aspect-ratio.html` test is checking. Changing this to ignore the attributes makes the test fail:

```
  ▶ TIMEOUT [expected OK] /html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio.html
  │   ▶ FAIL [expected PASS] Video width and height attributes are used to infer aspect-ratio
  │   │   → assert_approx_equals: expected 2.5 +/- 0.001 but got 2
  │   │ 
  │   │ assert_ratio@http://web-platform.test:8000/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio.html:19:23
  │   │ @http://web-platform.test:8000/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio.html:30:17
  │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2622:25
  │   │ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:738:36
  └   └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:737:20
```

Reply:
OK, so that happens because `width="100" height="100"` set `aspect-ratio: auto 100 / 100`, and then `auto` prefers the natural aspect ratio if available, which as per https://github.com/w3c/csswg-drafts/issues/7524#issuecomment-1204462924 it should never be from the default object size.

I added a commit to try to address this, let's see if it works